hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/api/DataNodeMetricsService.java

```diff
@@ -0,0 +1,349 @@
+      OzoneConfiguration config,
+      MetricsServiceProviderFactory metricsServiceProviderFactory) {
+    
+    this.reconNodeManager = (ReconNodeManager) reconSCM.getScmNodeManager();
+    this.httpsEnabled = HttpConfig.getHttpPolicy(config).isHttpsEnabled();
+    this.minimumApiDelayMs = (int) config.getTimeDuration(
+        OZONE_RECON_DN_METRICS_COLLECTION_MINIMUM_API_DELAY,
+        OZONE_RECON_DN_METRICS_COLLECTION_MINIMUM_API_DELAY_DEFAULT,
+        TimeUnit.MILLISECONDS);
+    this.maximumTaskTimeout = (int) config.getTimeDuration(OZONE_RECON_DN_METRICS_COLLECTION_TIMEOUT,
+        OZONE_RECON_DN_METRICS_COLLECTION_TIMEOUT_DEFAULT, TimeUnit.MILLISECONDS);
+    this.metricsServiceProviderFactory = metricsServiceProviderFactory;
+    this.lastCollectionEndTime.set(-minimumApiDelayMs);
+    int corePoolSize = Runtime.getRuntime().availableProcessors() * 2;
+    this.executorService = new ThreadPoolExecutor(
+        corePoolSize, MAX_POOL_SIZE,
+        KEEP_ALIVE_TIME, TimeUnit.SECONDS,
+        new LinkedBlockingQueue<>(),
+        new ThreadFactoryBuilder()
+            .setNameFormat("DataNodeMetricsCollector-%d")
+            .build());
+  }
+
+  /**
+   * Starts the metrics collection task if not already running and rate limit allows.
+   */
+  public void startTask() {
+    // Check if already running
+    if (!isRunning.compareAndSet(false, true)) {
+      LOG.warn("Metrics collection already in progress, skipping");
+      return;
+    }
+    
+    // Check rate limit
+    if (System.currentTimeMillis() - lastCollectionEndTime.get() < minimumApiDelayMs) {
+      LOG.debug("Rate limit active, skipping collection (delay: {}ms)", minimumApiDelayMs);
+      isRunning.set(false);
+      return;
+    }
+
+    Set<DatanodeDetails> nodes = reconNodeManager.getNodeStats().keySet();
+    if (nodes.isEmpty()) {
+      LOG.warn("No datanodes found to query");
+      resetState();
+      currentStatus = MetricCollectionStatus.FINISHED;
+      isRunning.set(false);
+      return;
+    }
+
+    // Set status immediately before starting async collection
+    currentStatus = MetricCollectionStatus.IN_PROGRESS;
+    LOG.debug("Starting metrics collection for {} datanodes", nodes.size());
+    
+    // Run a collection asynchronously so status can be queried
+    CompletableFuture.runAsync(() -> collectMetrics(nodes), executorService)
+        .exceptionally(throwable -> {
+          LOG.error("Metrics collection failed", throwable);
+          synchronized (DataNodeMetricsService.this) {
+            currentStatus = MetricCollectionStatus.FINISHED;
+            isRunning.set(false);
+          }
+          return null;
+        });
+  }
+
+  /**
+   * Collects metrics from all datanodes. Processes completed tasks first, waits for all.
+   */
+  private void collectMetrics(Set<DatanodeDetails> nodes) {
+    CollectionContext context = submitMetricsCollectionTasks(nodes);
+    processCollectionFutures(context);
+    updateFinalState(context);
```

Comment:
We should set result in finally, as if any unknown exception is thrown, still next iteration can continue